dotnet/Program.cs

```diff
@@ -1,10 +1,158 @@
+
+        /// <summary>
+        /// Calculate memory change with improved precision
+        /// </summary>
+        public long MemoryChange => EndMemory - StartMemory;
+    }
+
+    /// <summary>
+    /// Advanced memory profiler optimized for .NET 9
+    /// </summary>
+    private static class MemoryProfiler
+    {
+        /// <summary>
+        /// Get current process memory usage with .NET 9 improvements
+        /// </summary>
+        public static long GetMemoryUsage()
+        {
+            // Leverage .NET 9 cross-platform memory tracking
+            return GC.GetTotalMemory(false);
+        }
+    }
+
+    /// <summary>
+    /// Perform an asynchronous task with comprehensive tracking
+    /// </summary>
+    private static async Task<TaskMetrics> PerformTaskAsync(
+        int taskId, 
+        ILogger logger)
+    {
+        // Leverage .NET 9's improved task tracking
+        var startTime = DateTime.UtcNow;
+        var startMemory = MemoryProfiler.GetMemoryUsage();
+
+        try
+        {
+            // Simulate 10-second task with improved delay mechanism
+            await Task.Delay(TimeSpan.FromSeconds(10));
+
+            var endTime = DateTime.UtcNow;
+            var endMemory = MemoryProfiler.GetMemoryUsage();
+
+            // Advanced logging with structured logging
+            logger.LogInformation(
+                "Task {TaskId} completed. Duration: {Duration}ms, Memory Change: {MemoryChange} bytes", 
+                taskId, 
+                (endTime - startTime).TotalMilliseconds,
+                endMemory - startMemory
+            );
+
+            return new TaskMetrics(
+                TaskId: taskId, 
+                StartTime: startTime, 
+                EndTime: endTime, 
+                StartMemory: startMemory, 
+                EndMemory: endMemory
+            );
+        }
+        catch (Exception ex)
+        {
+            // Enhanced error handling with .NET 9 logging
+            logger.LogError(
+                ex, 
+                "Task {TaskId} failed with error: {ErrorMessage}", 
+                taskId, 
+                ex.Message
+            );
+
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Run benchmark with configurable task count
+    /// </summary>
+    private static async Task RunBenchmarkAsync(
+        int numTasks, 
+        ILogger logger)
+    {
+        // Leverage .NET 9's improved concurrent collections
+        var results = new ConcurrentBag<TaskMetrics>();
+
+        // Parallel task execution with improved task management
+        var tasks = Enumerable.Range(0, numTasks)
+            .Select(taskId => Task.Run(async () => 
+            {
+                var result = await PerformTaskAsync(taskId, logger);
+                results.Add(result);
+                return result;
+            }))
+            .ToList();
+
+        // Efficient task completion tracking
+        await Task.WhenAll(tasks);
```

Comment:
You are capturing locals in the closure and creating unnecessary delegates. 
Also, unlike Rust, by default .NET schedules async Tasks and yields immediately, so you don't need to wrap it in a `Task.Run` at all (which would result in creating two Tasks for every single async Task). And the usage to `ConcurrentBag` is unnecessary. 
The code should be simple to avoid introducing extra overhead which is unrelated to the concurrency structure itself.

```suggestion
        var tasks = new List<Task<TaskMetrics>>(numTasks);
        for (int i = 0; i < numTasks; i++)
        {
            tasks.Add(PerformTaskAsync(i, logger));
        }
        await Task.WhenAll(results); // this will gives you TaskMetrics[] which contains all the results
```